Change ConfigValue#get() to throw if called before config loaded

sciwhiz12 opened the pull request:
This PR implements and closes #7716 by changing the behavior of `ConfigValue#get` to throw if called before the config has been loaded. This prevents silent issues where a mod gets the value of the setting before configs are loaded, which means the default value is always returned.

As said in the linked RFC issue, there may be situations where the getting the config setting before configs are loaded is needed, and it is not preferable or maybe even possible to hardcode the default value. For these situations, the original behavior (return the actual value, or the default if config is not yet loaded) is made available through the new `#getRaw` method.

This PR also introduces javadocs for both `#get` and `#getRaw` to clarify the distinction between the two method's behaviors, and to point current users of `#get` which rely on the default-value-returning behavior to use the other method instead.

sciwhiz12 committed:
Change ConfigValue#get() to throw if called before config loaded

This prevents silent issues where a mod gets the value of the setting
before configs are loaded, which means the default value is always
returned.

As there may be situations where the getting the config setting before
configs are loaded is needed, and it is not preferable to hardcode the
default value, the original behavior is made available through #getRaw.

Implements and closes #7716

sciwhiz12 committed:
Make consistent the javadoc paragraph tags

sciwhiz12 committed:
Fix crash due to use of ConfigValue#get in patch

This call happens before client configs are loaded (as the
ClientModLoader#begin call happens after the call in question.)

malte0811 commented on patches/minecraft/net/minecraft/client/Minecraft.java.patch:
```diff
@@ -29,7 +29,7 @@
  
        Util.f_137440_ = RenderSystem.m_69576_();
-+      if (net.minecraftforge.common.ForgeConfig.CLIENT.forceSystemNanoTime.get()) Util.f_137440_ = System::nanoTime;
++      if (net.minecraftforge.common.ForgeConfig.CLIENT.forceSystemNanoTime.getRaw()) Util.f_137440_ = System::nanoTime;
```
If this runs before configs are loaded, doesn't that mean that the option does not do anything? This code only runs once during startup, the usual usecase for `getRaw` should be code that is called both before and after the config is loaded.

Thanks for implementing this btw!

sciwhiz12 replied:
Yes; as currently implemented, the `forceSystemNanoTime` option on the Forge client config does not affect anything at all. By this reasoning, the config option should be removed (or possibly moved to a later position), but I'm presuming that change would be out of this PR's scope, and am leaving it in for now for a later PR to remove it. 

If wanted, I can mark the option for removal in a later PR with a `// TODO` comment, or fix/remove the option in this PR, though I'd like to keep the scope of this PR as-is and leave the matter of fixing/removing it for someone else more knowledgable on its effects (or lack thereof) on rendering).

ichttt replied:
This is actually the perfect example for the use case of this PR - the current `get` method may cause cases where the config isn't respected without the modder recognising this. I would at least add a `TODO` comment here, but if you want to fix it directly in this PR I personally wouldn't consider this Out of Scope

sciwhiz12 replied:
I've marked it as a TODO for now, since this needs to be reevaluated more closely by people more knowledgable of `glfwGetTime` possibly having performance issues (see the commit message of 06804675be18c612138a713aca1a5801a7e5dbf8 which added this).

sciwhiz12 committed:
Merge upstream 1.18.x

sciwhiz12 committed:
Mark the use of getRaw in patch with a TODO

sciwhiz12 committed:
Update license header year in modified file

ichttt reviewed (approved):
Code LGTM. I don't know if it's too late in the 1.18.1 lifecycle to introduce this (even though it isn't binary breaking and we would even allow binary breaks as we don't have a RB yet), but it might cause some mods to crash on startup. I'll leave that up to the core team to decide.

sciwhiz12 committed:
Merge latest 1.18.x

sciwhiz12 committed:
Change check to dev-time only for now

This prevents the change from breaking already published mods, while
forcing mod developers to fix their incorrect usages in their
development environment.

sciwhiz12 commented:
After a bit of (internal) discussion, I decided to change the check to be dev-time only for now, with comments and an exception message notifying developers that this will be upgraded to a full check (both dev-time and production-time) in the future, such as during 1.19.

This is to reduce the breakage caused by this PR, by allowing already published mods to run without error, while erroring out in the dev environment so developers can fix their incorrect and untimely usages of `#get`.

sciwhiz12 committed:
Merge latest 1.18.x

sciwhiz12 committed:
Add getDefault() method to retrieve default value

sciwhiz12 committed:
Remove getRaw() method

This is effectively replaced with the expression `spec.isLoaded() ?
configValue.get() : configValue.getDefault()`.

sciwhiz12 commented:
After a bit more internal discussion, `getRaw()` has been removed and `getDefault()` added in its place. Any would-be calls to `getRaw()` should now be the expression `spec.isLoaded() ? configValue.get() : configValue.getDefault()` (where `ForgeConfigSpec spec` and `ConfigValue<?> configValue`).

sciwhiz12 committed:
Remove forceSystemNanoTime config setting

As implemented, it never had any effect as any place where the config
value would be queried happens before the configs are loaded.

sciwhiz12 committed:
Merge and update to latest 1.19.x

sciwhiz12 commented:
PR has been updated to 1.19.

TheCurle commented:
lgtm
